### PR TITLE
ENH: special: use from_dlpack for array conversions

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-21828

#### What does this implement/fix?
`xp.asarray` does not guarantee that other array types are accepted as input. Until recently, it happened to work for all libraries ~~as long as everything stayed on the CPU, but it does not allow for for device transfers, and~~ but `array-api-strict` would like to remove conversions via `__array__` (which was included for convenience despite not being in the standard), breaking conversion between NumPy and strict arrays.

This adds a function `xp_asarray` which is like `xp.asarray` but accepts:
- `xp` - the output array type
- ~~`device` - the output array device~~ no longer used
- `order` and `subok` (for NumPy only)

and uses `from_dlpack` to convert between array types.

As a proof of concept, the PR replaces use of `xp.asarray` with `xp_asarray(..., xp=xp)` in `_special.support_alternative_backends` and the `xp` testing functions. 

#### Additional information
No longer important to this PR because it doesn't use `device`:
<details>

Either I'm misunderstanding, or a lot of issues remain with device support and `from_dlpack`. For example:

- ~~NumPy `from_dlpack` only understands device='cpu' (not a device object)~~ *Update: I guess the standard doesn't specify what a device object should be.*
- CuPy `from_dlpack` doesn't  accept a `device` argument, and when calling `from_dlpack(array)`, it recommends "Use `cupy.array(numpy.from_dlpack(input))` instead."
- Torch `from_dlpack` doesn't seem to work with any library (not even torch to torch)
- ~~The format of the return of `device` seems different in different libraries.~~ *Update: I guess the standard says this is out of scope.* For example: `xp.__array_namespace_info__().default_device()` returns the following in different libraries:
  - NumPy: `"cpu"`
  - array-api-strict: `array_api_strict.Device('CPU_DEVICE')`
  - torch via array-api-compat: `device(type='cpu')`
  - jax.numpy: `None`
  - CuPy via array-api-compat: `<CUDA Device 0>`
- (*new!*) see array API CI failure

In the meantime, I've added workarounds as needed while ensuring that `from_dlpack` is used when converting to array-api-strict.

</details>

### CODE REVIEW

File: scipy/_lib/_array_api.py [hunk 1]
```diff
@@ -195,6 +196,109 @@ def _asarray(
...
+        return device
+
+    xp_in = array_namespace(array)
+
+    # Here is what the logic is intended to be
+    # if is_numpy(xp_out) or is_array_api_strict(xp_out) or is_cupy(xp_out):
+    #     # These libraries only support one type of device
+    #     return xp_out.__array_namespace_info__().default_device()
+    # else:
+    #     return xp_in.device(array)
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:210]: I suppose `None` would do the same thing?

File: scipy/_lib/_array_api.py [hunk 2]
```diff
@@ -195,6 +196,109 @@ def _asarray(
...
+        array = np.asanyarray(array)
+
+    # Start with special cases that shouldn't be needed, but are due to
+    # shortcomings in library implementations of `device` and `from_dlpack`
+    if is_cupy(xp) or is_torch(xp):
+        # Torch from_dlpack doesn't work with any library - not even torch to torch
+        # CuPy `from_dlpack` doesn't accept a device argument, and when calling
+        # without one, it recommends:
+        # Use `cupy.array(numpy.from_dlpack(input))` instead.
+        array = xp.asarray(np_compat.from_dlpack(array).copy(), dtype=dtype, copy=copy)
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:267]: `np_compat.from_dlpack(array)` is typically read-only, and `torch` doesn't support that as input. I probably shouldn't *always* copy, but I'm not sure if we need to split hairs about respecting `copy` until the ecosystem settles down a bit.

File: scipy/special/_support_alternative_backends.py [hunk 3]
```diff
@@ -41,9 +41,9 @@ def get_array_special_func(f_name, xp, n_array_args):
     def __f(*args, _f=_f, _xp=xp, **kwargs):
         array_args = args[:n_array_args]
         other_args = args[n_array_args:]
-        array_args = [np.asarray(arg) for arg in array_args]
+        array_args = [xp_asarray(arg, xp=np) for arg in array_args]
         out = _f(*array_args, *other_args, **kwargs)
-        return _xp.asarray(out)
+        return xp_asarray(out, xp=xp)
```
Comment (mdhaber) [inline scipy/special/_support_alternative_backends.py:46]: Outputs will typically be read-only after this step. This seems like a big disadvantage compared to `xp.asarray` using the array protocol. I don't think we want to return read-only results from special functions. Is there a general solution that is efficient (e.g. not just `copy` the output before returning it) or do we have to be very careful about using `from_dlpack` only when necessary?

File: scipy/_lib/_array_api.py [hunk 4]
```diff
@@ -195,6 +196,109 @@ def _asarray(
...
+        # I don't know if other libraries support `from_dlpack`, either
+        # We might need other special cases like this.
+    elif is_numpy(xp) and is_torch(xp_in):
+        # NumPy.from_dlpack doesn't work with Torch
+        # Otherwise: Tensor.__dlpack__() got an unexpected keyword argument 'dl_device'
+        array = xp.asarray(array, dtype=dtype, copy=copy)
+    elif is_numpy(xp_in) and not array.__class__.__name__.endswith('.ndarray') and subok:
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:274]:
```suggestion
    elif is_numpy(xp_in) and not array.__class__.__name__.endswith('ndarray') and subok:
```
For PEP8. Hopefully subclass names don't end with `ndarray`?

Or is there a better way to look for strict array subclasses?

File: scipy/_lib/_array_api.py [hunk 5]
```diff
@@ -34,8 +34,9 @@
     'get_xp_devices',
     'is_array_api_strict', 'is_complex', 'is_cupy', 'is_jax', 'is_numpy', 'is_torch', 
     'SCIPY_ARRAY_API', 'SCIPY_DEVICE', 'scipy_namespace_for',
+    'xp_asarray',
     'xp_assert_close', 'xp_assert_equal', 'xp_assert_less',
-    'xp_copy', 'xp_copysign', 'xp_device',
+    'xp_atleast_nd', 'xp_copy', 'xp_copysign', 'xp_device',
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:39]:
```suggestion
    'xp_copy', 'xp_copysign', 'xp_device',
```

File: scipy/_lib/_array_api.py [hunk 6]
```diff
@@ -199,6 +200,74 @@ def _asarray(
...
+    """Recursively expand the dimension to have at least `ndim`."""
+    if xp is None:
+        xp = array_namespace(x)
+    x = xp.asarray(x)
+    if x.ndim < ndim:
+        x = xp.expand_dims(x, axis=0)
+        x = xp_atleast_nd(x, ndim=ndim, xp=xp)
+    return x
+
+
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:270]:
```suggestion
```

**Review by mdhaber — COMMENTED** [review]
Merge mistake, I guess.

File: scipy/_lib/_array_api.py (hunk 4 above)
Comment (lucascolley) [reply scipy/_lib/_array_api.py]: You can just use `not (type(array) is np.ndarray)`, I'm pretty sure. `isinstance` accepts subclass relationships, but `type(...) is` does not.

**Comment by lucascolley** [comment]
I'm going to look at cleaning this up and extending to cover current usage of `__array__` for cross-library conversion. Let's come back to the question of read-only arrays once that is done.

**Pull request closed without merging**